File: python/teleia/store.py
from __future__ import annotations

import json
import os
import sqlite3
import time
import uuid
from pathlib import Path

from .llm import Message, ToolCall


def _data_path() -> Path:
    base = os.environ.get("XDG_DATA_HOME") or os.path.join(os.path.expanduser("~"), ".local", "share")
    return Path(base) / "teleia" / "teleia.sqlite"


class Store:
    def __init__(self) -> None:
        path = _data_path()
        path.parent.mkdir(parents=True, exist_ok=True)
        self.conn = sqlite3.connect(path)
        self.conn.executescript(
            """
            CREATE TABLE IF NOT EXISTS sessions (
                id TEXT PRIMARY KEY,
                model TEXT NOT NULL,
                created_at INTEGER NOT NULL
            );
            CREATE TABLE IF NOT EXISTS messages (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                session_id TEXT NOT NULL,
                seq INTEGER NOT NULL,
                payload TEXT NOT NULL
            );
            CREATE INDEX IF NOT EXISTS idx_messages_session ON messages(session_id, seq);
            CREATE TABLE IF NOT EXISTS aliases (
                name TEXT PRIMARY KEY,
                session_id TEXT NOT NULL,
                created_at INTEGER NOT NULL
            );
            """
        )
        self.conn.commit()

    def create_session(self, model: str) -> str:
        sid = uuid.uuid4().hex
        self.conn.execute(
            "INSERT INTO sessions (id, model, created_at) VALUES (?, ?, ?)",
            (sid, model, int(time.time())),
        )
        self.conn.commit()
        return sid

    def append(self, session_id: str, seq: int, message: Message) -> None:
        payload = json.dumps(_to_jsonable(message))
        self.conn.execute(
            "INSERT INTO messages (session_id, seq, payload) VALUES (?, ?, ?)",
            (session_id, seq, payload),
        )
        self.conn.commit()

    def load(self, session_id: str) -> list[Message]:
        rows = self.conn.execute(
            "SELECT payload FROM messages WHERE session_id = ? ORDER BY seq ASC",
            (session_id,),
        ).fetchall()
        return [_from_jsonable(json.loads(p)) for (p,) in rows]

    def save_alias(self, name: str, session_id: str) -> None:
        self.conn.execute(
            "INSERT OR REPLACE INTO aliases (name, session_id, created_at) VALUES (?, ?, ?)",
            (name, session_id, int(time.time())),
        )
        self.conn.commit()

    def resolve_alias(self, name: str) -> str:
        row = self.conn.execute(
            "SELECT session_id FROM aliases WHERE name = ?", (name,)
        ).fetchone()
        if row is None:
            raise RuntimeError(f"no session saved as '{name}'")
        return row[0]


def _to_jsonable(m: Message) -> dict:
    out: dict = {"role": m.role}
    if m.content is not None:
        out["content"] = m.content
    if m.tool_calls:
        out["tool_calls"] = [{"id": tc.id, "name": tc.name, "arguments": tc.arguments} for tc in m.tool_calls]
    if m.tool_call_id is not None:
        out["tool_call_id"] = m.tool_call_id
    return out


def _from_jsonable(raw: dict) -> Message:
    return Message(
        role=raw["role"],
        content=raw.get("content"),
        tool_calls=[ToolCall(id=tc["id"], name=tc["name"], arguments=tc["arguments"]) for tc in raw.get("tool_calls", [])],
        tool_call_id=raw.get("tool_call_id"),
    )


__all__ = ["Store", "Message", "ToolCall"]

File: lua/teleia/store.lua
-- Session/message persistence via shell-out to sqlite3.
local json = require("teleia.json")

local M = {}

local function shell_escape(s)
  return "'" .. tostring(s):gsub("'", "'\\''") .. "'"
end

local function data_path()
  local base = os.getenv("XDG_DATA_HOME")
  if not base or base == "" then
    local home = os.getenv("HOME") or "."
    base = home .. "/.local/share"
  end
  return base .. "/teleia/teleia.sqlite"
end

local function run_sqlite_write(path, sql)
  local tmp_sql = os.tmpname()
  local f = assert(io.open(tmp_sql, "w"))
  f:write(sql)
  f:close()
  os.execute(string.format("sqlite3 %s < %s > /dev/null 2>&1", shell_escape(path), shell_escape(tmp_sql)))
  os.remove(tmp_sql)
end

local function run_sqlite_query(path, sql)
  local tmp_sql = os.tmpname()
  local tmp_out = os.tmpname()
  local f = assert(io.open(tmp_sql, "w"))
  f:write(sql)
  f:close()
  os.execute(string.format("sqlite3 -separator '|' %s < %s > %s 2>/dev/null",
    shell_escape(path), shell_escape(tmp_sql), shell_escape(tmp_out)))
  local fo = assert(io.open(tmp_out, "r"))
  local out = fo:read("*a")
  fo:close()
  os.remove(tmp_sql)
  os.remove(tmp_out)
  return out
end

function M.open()
  local path = data_path()
  local dir = path:match("^(.*)/[^/]+$")
  if dir then os.execute("mkdir -p " .. shell_escape(dir)) end
  run_sqlite_write(path, [[
    CREATE TABLE IF NOT EXISTS sessions (
      id TEXT PRIMARY KEY,
      model TEXT NOT NULL,
      created_at INTEGER NOT NULL
    );
    CREATE TABLE IF NOT EXISTS messages (
      id INTEGER PRIMARY KEY AUTOINCREMENT,
      session_id TEXT NOT NULL,
      seq INTEGER NOT NULL,
      payload TEXT NOT NULL
    );
    CREATE INDEX IF NOT EXISTS idx_messages_session ON messages(session_id, seq);
    CREATE TABLE IF NOT EXISTS aliases (
      name TEXT PRIMARY KEY,
      session_id TEXT NOT NULL,
      created_at INTEGER NOT NULL
    );
  ]])
  return { path = path }
end

local function rand_id()
  math.randomseed(os.time() + math.floor(os.clock() * 1e6))
  local hex = "0123456789abcdef"
  local out = {}
  for i = 1, 32 do
    local idx = math.random(1, 16)
    out[i] = hex:sub(idx, idx)
  end
  return table.concat(out)
end

local function sql_quote(s)
  return "'" .. tostring(s):gsub("'", "''") .. "'"
end

function M.create_session(store, model)
  local id = rand_id()
  run_sqlite_write(store.path, string.format(
    "INSERT INTO sessions (id, model, created_at) VALUES (%s, %s, %d);",
    sql_quote(id), sql_quote(model), os.time()
  ))
  return id
end

function M.append(store, session_id, seq, message)
  local payload = json.encode(message)
  run_sqlite_write(store.path, string.format(
    "INSERT INTO messages (session_id, seq, payload) VALUES (%s, %d, %s);",
    sql_quote(session_id), seq, sql_quote(payload)
  ))
end

function M.load(store, session_id)
  local out = run_sqlite_query(store.path, string.format(
    "SELECT payload FROM messages WHERE session_id = %s ORDER BY seq ASC;",
    sql_quote(session_id)
  ))
  local msgs = {}
  for line in out:gmatch("([^\n]+)") do
    local ok, m = pcall(json.decode, line)
    if ok then table.insert(msgs, m) end
  end
  return msgs
end

function M.save_alias(store, name, session_id)
  run_sqlite_write(store.path, string.format(
    "INSERT OR REPLACE INTO aliases (name, session_id, created_at) VALUES (%s, %s, %d);",
    sql_quote(name), sql_quote(session_id), os.time()
  ))
end

function M.resolve_alias(store, name)
  local out = run_sqlite_query(store.path, string.format(
    "SELECT session_id FROM aliases WHERE name = %s;", sql_quote(name)
  ))
  out = out:gsub("%s+$", "")
  if out == "" then
    error("no session saved as '" .. name .. "'")
  end
  return out
end

return M
